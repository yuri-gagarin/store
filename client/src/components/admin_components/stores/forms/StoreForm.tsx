import React, { useState, useEffect, useRef }  from "react";
import { Button, Form, TextArea } from "semantic-ui-react";

type Props = {
  title: string;
  description: string;
  handleCreateStore(title: string, description: string): void;
  handleUpdateStore(title: string, description: string): void;
}
type FormState = {
  title: string;
  description: string;
}

const StoreForm: React.FC<Props> = ({ title, description, handleCreateStore, handleUpdateStore }): JSX.Element => {

  const [ newForm, setNewForm ] = useState<boolean>(true)
  const [ formState, setFormState ] = useState<FormState>({ title, description });
  const storeFormRef = useRef<HTMLDivElement>(document.createElement("div"));
  
  const handleTitleChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setFormState({
      ...formState,
      title: e.target.value
    });
  };
  const hadnleDescriptionChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      description: e.target.value
    });
  };
  const handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (newForm) {
      handleCreateStore(formState.title, formState.description);
    } else {
      handleUpdateStore(formState.title, formState.description);
    }
  };

  useEffect(() => {
    if (title && description) {
      setNewForm(false);
    } else {
      setNewForm(true);
    }
  },  [title, description]);

  useEffect(() => {
    if (storeFormRef.current) {
      const elem = storeFormRef.current.getBoundingClientRect();
      window.scrollTo({
        top: elem.bottom,
        behavior: "smooth"
      })
    }
  }, [storeFormRef]);
  
  return (
    <div id="createStoreFormHolder" ref={storeFormRef}>
      <Form id="createStoreForm">
        <Form.Field>
          <label>Store title</label>
          <input 
            onChange={handleTitleChange} 
            placeholder="Store title here ..." 
            value={formState.title}
          />
        </Form.Field>
        <Form.Field
          id='form-textarea-control-opinion'
          control={TextArea}
          label='Store Description'
          onChange={hadnleDescriptionChange}
          placeholder='Store description here...'
          value={formState.description}
         />
         {
           newForm 
            ? <Button type='submit' onClick={handleSubmit} content= "Create Store" />
            : <Button type='submit' onClick={handleSubmit} content= "Update Store" />
         }  
      </Form>
    </div>
  );
};

export default StoreForm;