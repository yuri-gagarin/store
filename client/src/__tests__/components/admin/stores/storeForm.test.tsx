import React from "react";
import StoreForm from "../../../../components/admin_components/stores/forms/StoreForm";
import { mount, ReactWrapper } from "enzyme";
// helpers //

describe("Store Form Component render tests", () => {
  describe("Empty Form render tests", () => {
    let wrapper: ReactWrapper;

    beforeAll(() => {
      window.scrollTo = jest.fn;
      // mock data //
      const storeData: IStoreData = {
        _id: "",
        title: "",
        description: "",
        images: [],
        createdAt: ""
      }
      // mount form //
      wrapper = mount(
        <StoreForm  
          title={storeData.title}
          description={storeData.description}
          newForm={true}
          handleCreateStore={jest.fn}
          handleUpdateStore={jest.fn}
        />
      );
    });

    it("Should properly mount", () => {
      expect(wrapper).toMatchSnapshot();
    });
    it("Should render '#storeFormTitleInput'", () => {
      expect(wrapper.render().find("#storeFormTitleInput").length).toEqual(1);
    });
    it("Should render '#storeFormDescInput'", () => {
      expect(wrapper.render().find("#storeFormDescInput").length).toEqual(1);
    });
    it("Should render '#adminStoreFormCreate' Button", () => {
      expect(wrapper.render().find("#adminStoreFormCreate").length).toEqual(1);
    });
    it("Should NOT render '#adminStoreFormUpdate' Button'", () => {
      expect(wrapper.render().find("#adminStoreFormUpdate").length).toEqual(0);
    });
  });
  
  describe("Form with valid data render tests", () => {
    let wrapper: ReactWrapper; let mockData: IStoreData;

    beforeAll(() => {
      window.scrollTo = jest.fn;
      // mock store data //
      mockData = {
        _id: "1",
        title: "title",
        description: "description",
        images: [],
        createdAt: "now"
      };

      wrapper = mount(
        <StoreForm 
          title={mockData.title}
          description={mockData.description}
          newForm={false}
          handleCreateStore={jest.fn}
          handleUpdateStore={jest.fn}
        />
      );
    });

    it("Should properly mount", () => {
      expect(wrapper).toMatchSnapshot();
    });
    it("Should render '#storeFormTitleInput'", () => {
      expect(wrapper.render().find("#storeFormTitleInput").length).toEqual(1);
    });
    it("Should set correct data in '#storeFormTitleInput'", () => {
      const input = wrapper.render().find("#storeFormTitleInput");
      expect(input.val()).toEqual(mockData.title);
    });
    it("Should render '#storeFormDescInput'", () => {
      expect(wrapper.render().find("#storeFormDescInput").length).toEqual(1);
    });
    it("Should set correct data in '#storeFormDescInput'", () => {
      const input = wrapper.render().find("#storeFormDescInput");
      expect(input.val()).toEqual(mockData.description);
    });
    it("Should render '#adminStoreFormUpdate' Button'", () => {
      expect(wrapper.render().find("#adminStoreFormUpdate").length).toEqual(1);
    });
    it("Should NOT render '#adminStoreFormCreate' Button", () => {
      expect(wrapper.render().find("#adminStoreFormCreate").length).toEqual(0);
    });
  });
});