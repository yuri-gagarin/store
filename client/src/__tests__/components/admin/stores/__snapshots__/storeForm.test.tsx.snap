// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store Form Component render tests Empty Form render tests Should properly mount 1`] = `
<StoreForm
  description=""
  handleCreateStore={[Function]}
  handleUpdateStore={[Function]}
  newForm={true}
  title=""
>
  <div
    id="createStoreFormHolder"
  >
    <Form
      as="form"
      id="createStoreForm"
    >
      <form
        className="ui form"
        id="createStoreForm"
        onSubmit={[Function]}
      >
        <FormField>
          <div
            className="field"
          >
            <label>
              Store title
            </label>
            <input
              id="storeFormTitleInput"
              onChange={[Function]}
              placeholder="Store title here ..."
              value=""
            />
          </div>
        </FormField>
        <FormField
          control={[Function]}
          id="storeFormDescInput"
          label="Store Description"
          onChange={[Function]}
          placeholder="Store description here..."
          value=""
        >
          <div
            className="field"
          >
            <label
              htmlFor="storeFormDescInput"
            >
              Store Description
            </label>
            <TextArea
              aria-describedby={null}
              as="textarea"
              id="storeFormDescInput"
              onChange={[Function]}
              placeholder="Store description here..."
              rows={3}
              value=""
            >
              <Ref
                innerRef={
                  Object {
                    "current": <textarea
                      id="storeFormDescInput"
                      placeholder="Store description here..."
                      rows="3"
                    />,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <textarea
                        id="storeFormDescInput"
                        placeholder="Store description here..."
                        rows="3"
                      />,
                    }
                  }
                >
                  <textarea
                    aria-describedby={null}
                    id="storeFormDescInput"
                    onChange={[Function]}
                    onInput={[Function]}
                    placeholder="Store description here..."
                    rows={3}
                    value=""
                  />
                </RefFindNode>
              </Ref>
            </TextArea>
          </div>
        </FormField>
        <Button
          as="button"
          content="Create Store"
          id="adminStoreFormCreate"
          onClick={[Function]}
          type="submit"
        >
          <Ref
            innerRef={
              Object {
                "current": <button
                  class="ui button"
                  id="adminStoreFormCreate"
                  type="submit"
                >
                  Create Store
                </button>,
              }
            }
          >
            <RefFindNode
              innerRef={
                Object {
                  "current": <button
                    class="ui button"
                    id="adminStoreFormCreate"
                    type="submit"
                  >
                    Create Store
                  </button>,
                }
              }
            >
              <button
                className="ui button"
                id="adminStoreFormCreate"
                onClick={[Function]}
                type="submit"
              >
                Create Store
              </button>
            </RefFindNode>
          </Ref>
        </Button>
      </form>
    </Form>
  </div>
</StoreForm>
`;

exports[`Store Form Component render tests Form with valid data render tests Should properly mount 1`] = `
<StoreForm
  description="description"
  handleCreateStore={[Function]}
  handleUpdateStore={[Function]}
  newForm={false}
  title="title"
>
  <div
    id="createStoreFormHolder"
  >
    <Form
      as="form"
      id="createStoreForm"
    >
      <form
        className="ui form"
        id="createStoreForm"
        onSubmit={[Function]}
      >
        <FormField>
          <div
            className="field"
          >
            <label>
              Store title
            </label>
            <input
              id="storeFormTitleInput"
              onChange={[Function]}
              placeholder="Store title here ..."
              value="title"
            />
          </div>
        </FormField>
        <FormField
          control={[Function]}
          id="storeFormDescInput"
          label="Store Description"
          onChange={[Function]}
          placeholder="Store description here..."
          value="description"
        >
          <div
            className="field"
          >
            <label
              htmlFor="storeFormDescInput"
            >
              Store Description
            </label>
            <TextArea
              aria-describedby={null}
              as="textarea"
              id="storeFormDescInput"
              onChange={[Function]}
              placeholder="Store description here..."
              rows={3}
              value="description"
            >
              <Ref
                innerRef={
                  Object {
                    "current": <textarea
                      id="storeFormDescInput"
                      placeholder="Store description here..."
                      rows="3"
                    >
                      description
                    </textarea>,
                  }
                }
              >
                <RefFindNode
                  innerRef={
                    Object {
                      "current": <textarea
                        id="storeFormDescInput"
                        placeholder="Store description here..."
                        rows="3"
                      >
                        description
                      </textarea>,
                    }
                  }
                >
                  <textarea
                    aria-describedby={null}
                    id="storeFormDescInput"
                    onChange={[Function]}
                    onInput={[Function]}
                    placeholder="Store description here..."
                    rows={3}
                    value="description"
                  />
                </RefFindNode>
              </Ref>
            </TextArea>
          </div>
        </FormField>
        <Button
          as="button"
          content="Update Store"
          id="adminStoreFormUpdate"
          onClick={[Function]}
          type="submit"
        >
          <Ref
            innerRef={
              Object {
                "current": <button
                  class="ui button"
                  id="adminStoreFormUpdate"
                  type="submit"
                >
                  Update Store
                </button>,
              }
            }
          >
            <RefFindNode
              innerRef={
                Object {
                  "current": <button
                    class="ui button"
                    id="adminStoreFormUpdate"
                    type="submit"
                  >
                    Update Store
                  </button>,
                }
              }
            >
              <button
                className="ui button"
                id="adminStoreFormUpdate"
                onClick={[Function]}
                type="submit"
              >
                Update Store
              </button>
            </RefFindNode>
          </Ref>
        </Button>
      </form>
    </Form>
  </div>
</StoreForm>
`;
